src/Analyzer/FunctionNode.cpp

```diff
@@ -25,25 +25,54 @@ FunctionNode::FunctionNode(String function_name_)
     children[arguments_child_index] = std::make_shared<ListNode>();
 }
 
-void FunctionNode::resolveAsFunction(FunctionOverloadResolverPtr function_value, DataTypePtr result_type_value)
+ColumnsWithTypeAndName FunctionNode::getArgumentTypes() const
 {
-    aggregate_function = nullptr;
+    ColumnsWithTypeAndName argument_types;
+    for (const auto & arg : getArguments().getNodes())
+    {
+        ColumnWithTypeAndName argument;
+        argument.type = arg->getResultType();
+        argument_types.push_back(argument);
+    }
+    return argument_types;
+}
+
+
+FunctionBasePtr FunctionNode::getFunction() const
+{
+    return std::dynamic_pointer_cast<IFunctionBase>(function);
+}
+
+AggregateFunctionPtr FunctionNode::getAggregateFunction() const
+{
+    return std::dynamic_pointer_cast<IAggregateFunction>(function);
+}
+
+bool FunctionNode::isAggregateFunction() const
+{
+    return typeid_cast<AggregateFunctionPtr>(function) != nullptr && !isWindowFunction();
```

Comment:
1. this version of typeid_cast creates a shard_ptr which is not needed; I would prefer `typeid_cast<const IAggregateFunction *>(function.get())`
2. IAggregateFunction is a base class, and to me looks like `typeid(*function) == typeid(IAggregateFunction)` is never true. I don't understand how it can work.
3. maybe add `isAggregateFunction` to `IResolvedFunction` ?

Reply:
I still don't understand how it works :D